### ISSUE TITLE
Post-Escopo 5 — reuse remaining allocations: sampler/, causal_attention scores, x_flat

### ISSUE THREAD

**Issue by fabious054** [fabious054/CandleCLI#16]
## Context

Final memory allocation/copy audit before Escopo 6 (CPU/`rayon` calibration) — re-reading the whole hot path fresh, not trusting earlier passes' memory. Found three real gaps, none discovered during Escopo 5's original audits (which focused on the KV cache and matmul kernel).

Initially triaged these as "too small to matter" based on absolute magnitude vs. `lm_head`'s cost. That triage was wrong — the deciding criterion for this project has never been magnitude, it's whether we're leaving our *own already-from-scratch* code with unreused allocations instead of building the same kind of reused-buffer mechanism already applied everywhere else (`RawKvCache`, `Q8_0Matmul`). All three qualify by that standard: none of them involve depending on a ready-made external abstraction — they're gaps in code we already wrote ourselves.

## 1. `sampler/` — never covered by any prior audit

`softmax_scaled` (`sampler/mod.rs`) allocates a fresh `Vec<f32>` sized to its input every call:
- `Temperature::sample` (the **default** strategy) calls it with the full ~152k-token vocabulary, every decode step.
- `TopP::sample` also calls it with the full vocabulary every step.
- `TopK::sample` only hits this path in the degenerate `k >= vocab` case.

`TopK::sample`'s `top_k_indices` still allocates `idx: Vec<usize> = (0..logits.len()).collect()` — a full ~152k-element vocabulary-sized `Vec`, every call, even though Escopo 4's Otimização 3B already removed the *sort* cost via `select_nth_unstable_by`. The allocation was never addressed, only the sorting algorithm.

`TopP::sample` has the same `idx` allocation, plus two more (`kept` collected, then re-mapped for normalization) — the most allocation-heavy of the three, on top of still doing a full sort (which is structurally necessary there — TopP needs sorted order to find the cumulative-probability cutoff, a data-dependent position, not a fixed `k`; this issue is about the allocations around it, not the sort itself).

## 2. `scores` in `causal_attention` (`kernel/attention.rs`)

```rust
pub fn causal_attention(...) -> Vec<f32> {
    ...
    let mut scores: Vec<f32> = Vec::new();  // fresh every call
    for i in 0..new_len {
        for h in 0..num_heads {
            scores.clear();
            scores.reserve(visible);  // visible grows with kv_len
            ...
        }
    }
}
```

`causal_attention` is a free function with no persistent owner, so `scores` starts empty on every call and must `reserve` up to `visible` (which, during decode, equals the full current `kv_len`) — every layer, every token. Unlike `RawKvCache`, this buffer never survives between calls, so the conversation-length-dependent grow cost repeats every single token instead of being paid once and then reused.

## 3. `x_flat` in `Attention::forward` / `Mlp::forward` (`arch/qwen/attention.rs`, `arch/qwen/mlp.rs`)

```rust
let x_flat = x.to_dtype(DType::F32)?.contiguous()?.flatten_all()?.to_vec1::<f32>()?;
```

`Tensor::to_vec1` always allocates a fresh `Vec`. Researched candle's actual API surface before proposing anything (per the project's standing "trace before proposing" practice): candle has no borrowed/strided CPU tensor view type (confirmed via the `dlpark` crate's own documentation of this exact limitation), so true zero-copy extraction isn't possible. But a lower-level path exists and is used by at least one other real crate (`kalosm-common`'s `copy_tensor_into_vec`): `Tensor::storage_and_layout()` → match `Storage::Cpu(cpu_storage)` → `cpu_storage.as_slice()?` → `contiguous_offsets()` → `into.clear(); into.extend_from_slice(&data[o1..o2])` into a **caller-owned, reused** `Vec`, instead of `to_vec1()`'s always-fresh allocation.

## What to implement

- **`sampler/`**: `Temperature`, `TopK`, `TopP` each own a reused `probs`/`idx`/`kept` scratch (`RefCell<Vec<_>>`, consistent with the `RefCell`-not-`Mutex` reasoning already established — samplers are called sequentially, never concurrently). `softmax_scaled` gains an `_into` variant mirroring `raw_forward_into`/`fused_swiglu_into`'s pattern; `top_k_indices` similarly writes into a reused `idx` buffer instead of collecting fresh.
- **`causal_attention`**: gains a `scores: &mut Vec<f32>` parameter instead of declaring it locally; `Attention` owns the persistent buffer (`scores_scratch: RefCell<Vec<f32>>`) and passes it in, same ownership pattern as `q_scratch`/`k_scratch`/`v_scratch`.
- **`x_flat`**: `Attention`/`Mlp` gain an `x_scratch: RefCell<Vec<f32>>`; replace `to_vec1()` with the `storage_and_layout()`-based copy into that reused buffer.

## Acceptance criteria

- [ ] No behavior change anywhere — every change here is pure reorganization of memory management, same arithmetic
- [ ] `cargo test --lib` green, no precision divergence
- [ ] New tests: `softmax_scaled`/`top_k_indices` into-variants match their existing Vec-returning counterparts; `causal_attention` with a pre-populated stale `scores` buffer (grown large from a prior call, then reused for a smaller one) matches a fresh-buffer call; `x_flat` extraction via the new path matches `to_vec1()` byte-for-byte on both contiguous and (if reachable in practice) non-contiguous input
- [ ] Existing tests unaffected: `attention_matches_oracle_equivalent`, `mlp_matches_oracle_activation`, `model_matches_oracle_end_to_end`, all sampler tests

## Expectation on impact (set explicitly, so results aren't misread)

Not expected to move the `▸ N.N tok/s` figure measurably — these allocations are small relative to `lm_head`/the matmul kernels, likely within the noise already seen between runs. The value is fewer allocator calls and reduced fragmentation risk over long conversations, and closing a real gap in code that's already fully our own — consistent with the project's standing principle that small, compounding improvements matter at scale (see `melhorias-pequenas-refletem-em-escala`, lobe atlas, CandleCLI satellite), not with any expectation of a visible number.

**Comment by fabious054** [comment]
## Implemented — verified before writing, not after

Branch `escopo5-followup/reuse-remaining-allocations`, 10 commits.

### 1. `sampler/` (never audited before)

- `sampler/mod.rs`: `softmax_scaled_into` writes into a caller-reused buffer instead of `softmax_scaled`'s always-fresh allocation (kept as a thin wrapper for tests/callers that want an owned result).
- `Temperature` (the **default** strategy): owns `probs_scratch`.
- `TopK`: owns `idx_scratch`/`kept_scratch`/`probs_scratch`. `top_k_indices_into` writes into a reused `idx` instead of `(0..logits.len()).collect()` every call — Otimização 3B (Escopo 4) removed the sort, this removes the allocation Otimização 3B left unaddressed.
- `TopP`: owns `probs_scratch`/`idx_scratch`/`kept_scratch`, and now normalizes `kept` in place instead of allocating it twice (collected, then re-collected during normalization) — a real bug-adjacent waste beyond just "not reused across calls."

### 2. `causal_attention`'s `scores` (`kernel/attention.rs`)

Was a local `Vec::new()` inside a free function with no persistent owner — started empty and grew back up to the current `kv_len` on *every single call*, every layer, every token. Now takes `scores: &mut Vec<f32>`; `Attention` owns `scores_scratch` and passes it in, same shape as `q_scratch`/`k_scratch`/`v_scratch`.

### 3. `x_flat` (`Attention::forward`, `Mlp::forward`) — verified by compiling, not assumed

`Tensor::to_vec1` always allocates. Before writing anything, set up an isolated Cargo project in this session's sandbox depending directly on the real `candle-core` (git clone, local path dependency — worked around a libgit2/proxy issue and an rustc/edition mismatch by installing newer toolchain packages via `apt`), and compiled + ran the exact API surface being proposed: `Tensor::storage_and_layout()` → `Storage::Cpu` → `CpuStorage::as_slice::<f32>()` → `Layout::contiguous_offsets()` → `extend_from_slice` into a reused `Vec`. Tested three cases directly: contiguous input, non-contiguous input (the `strided_index` fallback), and buffer reuse across a shrinking tensor size. All passed before any of this went into the real codebase. New `kernel/tensor_io.rs::tensor_into_vec`, used in both `Attention::forward` and `Mlp::forward`.

### The triage correction, for the record

First pass dismissed all three as "too small to matter" by comparing absolute allocation size against `lm_head`'s cost. That was wrong — the deciding criterion here has never been magnitude, it's whether a gap exists in code that's already entirely ours. Corrected mid-conversation; see `melhorias-pequenas-refletem-em-escala` (lobe atlas, CandleCLI satellite) for the fuller reasoning, now also captured as `DEVELOPMENT.md`/`CLAUDE.md` point 14.

### Validation

- No behavior change anywhere — pure reorganization, same arithmetic throughout.
- New tests: `softmax_scaled_into_reuse_matches_fresh`, `top_k_indices_into_reuse_matches_fresh`, `reused_scratch_survives_a_vocabulary_size_change` (TopP), `scores_scratch_reuse_does_not_leak_stale_data`, plus `tensor_io.rs`'s own three tests (contiguous, non-contiguous, reuse-drops-stale-data).
- Existing tests extended rather than just left passing: `attention_matches_oracle_equivalent`'s two-call structure now exercises all five of `Attention`'s scratch buffers (x/q/k/v/scores) shrinking back down, not just q/k/v as before; `mlp_matches_oracle_activation` similarly for `Mlp`'s x/gated.
- `causal_attention_matches_tensor_reference` updated to reuse the same `scores_scratch` across its prefill+decode calls, matching real usage.

### To run

```
git fetch origin
git checkout escopo5-followup/reuse-remaining-allocations
git pull origin escopo5-followup/reuse-remaining-allocations
cargo test --lib
```

Expect no `▸ N.N tok/s` movement worth reading into — flagged this explicitly going in, see the issue description's "Expectation on impact" section.
